---
title: "DAMEfinder Workflow"
author: 
- name: Dania Machlab
  email: dania.machlab@gmail.com
- name: Mark Robinson
  email: mark.robinson@imls.uzh.ch
package: DAMEfinder
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{DAMEfinder Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8} 

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```


# DAMEfinder Overview

`DAMEfinder` is a tool that detects differential allele-specifically methylated regions between conditions across samples using methylation counts of CpG tuples. Each tuple is assigned with an allele-specific methylation (ASM) score. For each CpG pair (or tuple) a moderated t-statistic is caluclated across all samples using the ASM scores. `limma`'s `lmFit` and `eBayes` functions are used in this process. The result is a t-statistic for each tuple. These values are used to detect the differentially allele-specifically methylated regions (DAMEs) with "bump hunting".  


## Bumphunting

# Prerequisites

The differential allele-specific methylation analysis using DAMEfinder starts at the level of the methtuple files. The user must have obtained a list of the files resluting from the methtuple tool to input them in the `read_tuples` function.

## Illustrative code on how to get from fastq files to the methtuple files level

The `DAMEfinder` package can be used for detection of differentially allele-specifically methylated regions starting from the files given by the methtuple tool. These files contain unique CpG tuple counts for each sample. We demonstrate how to get to this point starting from fastq files of bisulfite treated reads. Whether the BS-seq reads are genome-wide or targetting specific regions, the analysis is the same with `DAMEfinder`.

### Map bisulfite reads

? mention other tools for mapping BSseq reads.

We use `bismark` to map the BS-seq reads to a reference genome. Any trimming of the reads should be done before this step. In the example below, we map the PE reads of sample1 and show other the use of other helpful functions in bismark to exctract  methylation information that may be used later in the analysis. 
The path to the reference genome folder is "/home/Shared_taupo/data/annotation/_Archive/Human/genome/GRCH37" in this example. Because we have PE reads, we input two file names.

```{bash eval=FALSE}
# Map BS-seq reads
path_to_genome_folder="/home/Shared_taupo/data/annotation/_Archive/Human/genome/GRCH37"
bismark --bowtie2 -p 4 $path_to_genome_folder -1 sample1_R1.fastq.gz -2 sample1_R2.fastq.gz

# Remove duplicates
deduplicate_bismark  -p sample1_R1.fastq_bismark_bt2_pe.sam

# Get methylation information using bismark
bismark_methylation_extractor -p --comprehensive sample1_R1.fastq_bismark_bt2_pe.sam
bismark2bedGraph --counts -o CpG_context_sample1_R1.fastq_bismark_bt2_pe.bedGraph CpG_context_sample1_R1.fastq_bismark_bt2_pe.txt

# Convert sam files to bam files
samtools view -Sb sample1_R1.fastq_bismark_bt2_pe.sam > sample1_R1.fastq_bismark_bt2_pe.bam
samtools sort sample1_R1.fastq_bismark_bt2_pe.bam sample1_R1.fastq_bismark_bt2_pe_sorted
samtools index sample1_R1.fastq_bismark_bt2_pe_sorted.bam

```


### Run `methtuple` on bam files

After mapping the BS-seq reads and obtaining the resulting bam files, we can run `methtuple`. `methtuple` requires the input bam files of PE reads to be sorted by query name. For more information on the options in `methtuple`, refer to the manual of the software. There is the option, for example, of sckipping a number of bases from either end of each read. This is usefult to avoid the first few bases that show methylation bias. The --sc option combines strand information. The "tsv.gz" file produced contains the methylation counts for all the unique tuples of size 2 for sample1.


```{bash eval=FALSE}

# Sort bam files by query
samtools sort -n -@ 10 -m 20G -O bam -T _tmp -o QS_sample1_R1.fastq_bismark_bt2_pe_sorted.bam sample1_R1.fastq_bismark_bt2_pe_sorted.bam

# Run methtuple
methtuple --sc --gzip -m 2 QS_sample1_R1.fastq_bismark_bt2_pe_sorted.bam 

```



# Running `DAMEfinder`: Example Workflow 

We use an example data set found in the `DAMEfinder` package to illustrate the workflow. We copy all the files available in the extdata folder of the package which contains the `methtuple` files we will use in this example. The files are the result of simulated data and encompass a small section of chr14 and contain allele-scpecifically methylated regions in the normal samples, some of which are lost in the adenoma samples.


## Read in `methtuple` files 

We copy the example data set into the current working directory. The meta data file contains the file names, sample names and sample conditions. The methtuple files are read using the read_tuples function from `DAMEfinder`.



```{r}
library(DAMEfinder)

# Copy example data to workig directory
file.copy(system.file(package="DAMEfinder", "extdata"), ".", recursive=TRUE)
# Read in meta data file
meta_data <- read.table("extdata/metadata.txt", header = TRUE)
file_list <- list.files(path = "extdata/", pattern = ".tsv.gz")
m <- match(file_list, meta_data$methtuplefile)
file_list <- paste0("extdata/", file_list)
names(file_list) <- meta_data$shortname[m]
groups <- meta_data$group[m]

file_list

# Read in methtuple files
sample_list <- read_tuples(files = file_list, sample_names = names(file_list))

head(sample_list$"adenoma1")


```


The 'read_tuples' function returns a list of data frames, with each data frame corresponding to one sample. Each row in the data frame displays a tuple. The chromosome name and starnd are shown followed by pos1 and pos2 which refer to the genomic positions of the first and second CpG in the tuple. The MM, MU, UM, and UU counts of the tuple are displayed where M stands for methylated and U for unmethylated. For example, UM shows the read counts for the instances where pos1 is unmethylated and pos2 is methylated. The total read count (or coverage) and distance between the two genomic positions in the tuple are shown under "cov" and "inter_dist" respectively.

### Remove SNPs

On a real data set, tuples that contain CpGs corresponding to SNPs ought to be removed using the 'remove_snps' function. Since the example data set is a simulated one, there is no need to apply this step in our workflow. We demonstrate the use of the function below. A snp_key must be supplied as input, which contains all the SNPs in the following form: 'chr.position'. For example, "chr5.7382" is a SNP in chromosome 5 at position 7382.

suggestiong for future: specifiy the genome only and have the function produce the snp_key

```{r eval=FALSE}

sample_list_no_snps <- parallel::mclapply(sample_list, remove_snps, mc.cores=6)
```


## Calculate ASM Score

The 'calc_asm' function takes in the list of samples and outputs a `SummarizedExperiment` object, with an 'asm' assay that is a matrix where each row is a tuple and each column is a sample. This matrix contains allele-specific methylation (ASM) scores. The ASM score is a measure of the extent of allele-specific methylation. The higher the ASM score, the more evidence there is for allele-scpecific methylation for a particular tuple. Equations \@ref(eq:asmGeneral), \@ref(eq:asmWeight) and \@ref(eq:asmTheta) show how the score is calculated. The log odds ratio in equation \@ref(eq:asmGeneral) provides a higher score the more MM and UU counts the tuple has, rather than random methylations of UM and MU. The weight further adds a measure of allele-scpecific methylation where a rather balanced MM:UU count provides the tuple with a higher score. Adding this weight is necessary, since we can still have high log odds ratios with instances of fully methylated tuples (high MM count) and fully unmethylated ones (high UU count).

\begin{equation} 
    ASM^{(i)} = \log{ \Big\{ \frac{X_{MM}^{(i)} \cdot X_{UU}^{(i)}}{X_{MU}^{(i)} \cdot X_{UM}^{(i)}} \Big\} \cdot w_i }
    (\#eq:asmGeneral)
\end{equation}

\begin{equation}
    w_i = P(0.5-\epsilon < \theta < 0.5+\epsilon \mid X_{MM}^{(i)}, X_{UU}^{(i)}, \beta_1, \beta_2)
    (\#eq:asmWeight)
\end{equation}

\begin{equation}
    \theta^{(i)} \mid X_{MM}^{(i)}, X_{UU}^{(i)},\beta_1, \beta_2 \sim \operatorname{Beta}(\beta_1+X_{MM}^{(i)}, \beta_2+X_{UU}^{(i)})
    (\#eq:asmTheta)
\end{equation}


where $\theta^{(i)}$ represents the moderated proportion of MM to MM+UU alleles. The weight, $w_i$ is set such that the observed split between MM and UU alleles can depart somewhat from 50/50, while fully methylated or unmethylated tuples, which represents evidence for absence of allele-specificity, are attenuated to 0.  The degree of allowed departure can be set according to $\epsilon$, the deviation from 50/50 allowed and the level of moderation, $\beta_1$ and $\beta_2$.  For example, the degree of mixture of subpopulations may play a role (e.g., allele-specificity may not happen on all subpopulations).




```{r}

ASM_score_matrix <- calc_asm(sample_list = sample_list)

```


```{r}

ASM_score_matrix

head(SummarizedExperiment::assays(ASM_score_matrix)$asm)

```



At this point, the user may want to remove rows (tuples) in the ASM_score_matrix, where all samples of one condition have NA values. One can also set a minimum number of samples to have a specific count per tuple as a further restriction. However, removing tuples that have NA values in all samples of a particular condition is a necessary step.

```{r}

# Remove rows where all samples have NA 
asm <- SummarizedExperiment::assays(ASM_score_matrix)$asm
ind <- apply(asm, 1, function(x) all(is.na(x)))
ASM_score_matrix <- ASM_score_matrix[ !ind, ]

ASM_score_matrix

```



## Get t-Statistics per tuple

Next, we obtain a moderated t-statistic per tuple that reflects a measure of difference between the two conditions being comapred. The t-Statistic is smoothed using the 'smoother' function from `bumphunter`. The midpoint of the two positions in each tuple is set as the genomic position of every tuple in the smoothing process.

```{r, message=FALSE}

# Make a design matrix that specifies the two conditions per sample
cols <- rep(1,ncol(ASM_score_matrix))
n <- grep("normal", colnames(ASM_score_matrix))
cols[n] <- 0

mod <- matrix(data=c(rep(1,ncol(ASM_score_matrix)), cols), ncol = 2)
mod

# Get t-Statistics
tStatistics <- get_tstats(ASM_score_matrix, mod, method = "ls")

```

## Find DAMEs

Finally, we detect regions that show differential allele-specific methylation (DAMEs). The 'regionFinder' function from `bumphunter` is used in this process.

```{r}

dames <- find_dames(tStatistics)

head(dames)

```

The 'find_dames' function returns a data frame with each row corresponding to a DAME. The position of each DAME is shown in the "chr", "start" and "end" columns. The "area" column reflects the extent of differential allele-specific methylation and the DAMEs are ranked by this value.


# Session Information

```{r}
sessionInfo()
```



<!-- ## R Markdown -->

<!-- This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>. -->

<!-- When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this: -->

<!-- ```{r cars} -->
<!-- summary(cars) -->
<!-- ``` -->

<!-- ## Including Plots -->

<!-- You can also embed plots, for example: -->

<!-- ```{r pressure, echo=FALSE} -->
<!-- plot(pressure) -->
<!-- ``` -->

<!-- Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot. -->
